services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithUnknownContentLengthHelper.java

```diff
@@ -174,11 +174,13 @@ public void onNext(AsyncRequestBody asyncRequestBody) {
                         subscription.cancel();
                     } else {
                         uploadId = createMultipartUploadResponse.uploadId();
-                        uploadIdFuture.complete(uploadId);
                         log.debug(() -> "Initiated a new multipart upload, uploadId: " + uploadId);
 
                         sendUploadPartRequest(uploadId, firstRequestBody);
                         sendUploadPartRequest(uploadId, asyncRequestBody);
+
+                        // We need to complete the uploadIdFuture *after* the first two requests have been sent
+                        uploadIdFuture.complete(uploadId);
```

Comment:
The race condition is that the third request on line 189 could be executed before the first and second request, so the fix is completing `uploadIdFuture` *after* the first two requests have been finished.